<a href="https://colab.research.google.com/github/kahiga2244/Week-9-project/blob/master/Moringa_Core_wk9_2_Joseph_Kahiga.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1.Defining the question

a) Specifying the Data Analytic Question

For dataset 2, perform classification of the testing set samples using the Naive Bayes Classifier.
Compute the accuracy (percentage of correct classification).

b. Metric of success

Compute the accuracy (percentage of correct classification).
Report the confusion matrix of each classifier.
Repeat step 2 to step 4 twice, each time splitting the datasets differently i.e. 70-30, 60-40, then note the outcomes of your modeling.

c) Understanding the context

d) Recording the Experimental Design

e) Data Relevance

2. Loading the libraries

In [121]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import BernoulliNB

%matplotlib inline


b. loading the dataset

In [96]:
df = pd.read_csv('/content/spambase_csv.csv')

3. checking the data

In [97]:
df.head()

word_freq_make  word_freq_address  ...  capital_run_length_total  class
0            0.00               0.64  ...                       278      1
1            0.21               0.28  ...                      1028      1
2            0.06               0.00  ...                      2259      1
3            0.00               0.00  ...                       191      1
4            0.00               0.00  ...                       191      1

[5 rows x 58 columns]

In [98]:
df.tail()

word_freq_make  word_freq_address  ...  capital_run_length_total  class
4596            0.31                0.0  ...                        88      0
4597            0.00                0.0  ...                        14      0
4598            0.30                0.0  ...                       118      0
4599            0.96                0.0  ...                        78      0
4600            0.00                0.0  ...                        40      0

[5 rows x 58 columns]

In [99]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

In [100]:
df.isna().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

4. Tidying the data

In [101]:
df.shape

(4601, 58)

In [102]:
df.duplicated(subset=None, keep='first')

0       False
1       False
2       False
3       False
4       False
        ...  
4596    False
4597    False
4598    False
4599    False
4600    False
Length: 4601, dtype: bool

Outlier detection

In [103]:
# Create detector
from sklearn.covariance import EllipticEnvelope
outlier_detector = EllipticEnvelope(contamination=.1)

# Fit detector
outlier_detector.fit(df)

# Predict outliers
outlier_detector.predict(df)

/usr/local/lib/python3.6/dist-packages/sklearn/covariance/_robust_covariance.py:170: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-248.843085999315491 > -266.705553285045994). You may want to try with a higher value of support_fraction (current value: 0.507).
  RuntimeWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/covariance/_robust_covariance.py:170: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-205.740814489980551 > -260.029841431920886). You may want to try with a higher value of support_fraction (current value: 0.507).
  RuntimeWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/covariance/_robust_covariance.py:170: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-211.112090121072328 > -259.110220934536642). You may want to try with a higher value of support_fraction (current value: 0.507).
  RuntimeWarning)
/usr/local

array([1, 1, 1, ..., 1, 1, 1])

Gaussian Naive Bayes classifier

In [106]:
# Load libraries and datasets to be used in this example
#
from sklearn import datasets
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split

In [107]:
# Splitting our dataset into its attributes and labels
# ---
# The X variable contains the first four columns of the dataset (i.e. attributes) while y contains the labels.
# ---
# 
X = df.iloc[:, :-1].values
y = df.iloc[:, 53].values

In [108]:
#train test split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=4)

In [109]:
# Instantiate: create object
gnb = GaussianNB()

# Fit
gnb.fit(X_train, y_train)

# Predict
y_pred = gnb.predict(X_test)

# Accuracy
acc = accuracy_score(y_test, y_pred)
acc

ValueError: ignored

Implementing the solution

Multinomial Naive Bayes Classifier

In [110]:
#splitting
train_x, test_x, train_y, test_y = train_test_split(df.drop('class', axis=1),
                                                    df['class'],
                                                    test_size = 0.4, random_state=6)

In [114]:
#fitting our modelling

model = MultinomialNB().fit(train_x, train_y)

In [118]:
#evaluating our model
predictions = model.predict(test_x)
sum(predictions) / len(predictions)

0.40901683867463334

In [129]:
#accuracy
accuracy_rate = accuracy_score(test_y, predictions)
accuracy_rate

0.8104291146116241

changing the columns and sample size to 57

In [134]:
X = df.iloc[:, :-5].values
y = df.iloc[:, 45].values

In [135]:
#splitting
train_x, test_x, train_y, test_y = train_test_split(df.drop('class', axis=1),
                                                    df['class'],
                                                    test_size = 0.57, random_state=6)

In [140]:
# Instantiate: create object
#fitting our modelling

model = MultinomialNB().fit(train_x, train_y)


#evaluating our model
predictions = model.predict(test_x)
sum(predictions) / len(predictions)


0.4135159273593331

In [141]:
#accuracy
accuracy_rate = accuracy_score(test_y, predictions)
accuracy_rate

0.8154212563262876

changing the sample size to 73

In [163]:
X = df.iloc[:, 48].values
y = df.iloc[:, 48].values

In [164]:
#splitting
train_x, test_x, train_y, test_y = train_test_split(df.drop('class', axis=1),
                                                    df['class'],
                                                    test_size = 0.73, random_state=6)

In [165]:
# Instantiate: create object
#fitting our modelling

model = MultinomialNB().fit(train_x, train_y)


#evaluating our model
predictions = model.predict(test_x)
sum(predictions) / len(predictions)

0.4135159273593331

In [154]:
#accuracy
accuracy_rate = accuracy_score(test_y, predictions)
accuracy_rate

0.8154212563262876

Challenging the soultion

In [171]:
test_x = test_x[test_x.columns[:30]]
train_x = train_x[train_x.columns[:30]]

# Instantiate: create object
#fitting our modelling

model = MultinomialNB().fit(train_x, train_y)


#evaluating our model
predictions = model.predict(test_x)
#accuracy
accuracy_rate = accuracy_score(test_y, predictions)
accuracy_rate


0.808573980351295

Conclusion

The accuracy is at its at 0.8154212563262876

Gaussian naive bayes and bernoulli did not work so i had to disregard them and only use multinomial

changing the sample size changed the prediction results where more samples used gave better prediction
